src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -1010,8 +1010,12 @@ void ShenandoahFreeSet::flip_to_old_gc(ShenandoahHeapRegion* r) {
   size_t region_capacity = alloc_capacity(r);
   _free_sets.move_to_set(idx, OldCollector, region_capacity);
   _free_sets.assert_bounds();
-  _heap->generation_sizer()->force_transfer_to_old(1);
+
```

Comment:
IIUC, the new code will fail to transfer if the transfer makes YOUNG smaller than min_capacity or makes OLD larger than max_capacity.  It will never fail because unaffiliated_regions < 1.

I have two concerns with the code as written:
1. I don't think we should flip the region if we don't flip the capacity.  That will create future assertion failures.
2. If we fail to flip_to_old(), we need to return a status code to so indicate.  The caller needs to know that the requested flip cannot be achieved and thus the precipitating "allocation request" must fail.

Alternatively, we could adopt the convention that min/max generation sizes are "preferences" but are not absolutely enforced.  Maybe it is better to bend the rules to allow pause-free operation rather than forcing an old evacuation failure that would require a Full GC.

Reply:
If we adopt this convention, we should get rid of any asserts that endeavor to enforce max/min generation capacities  (or change these to enforce that capacity never shrinks below 0 and never grows above heap->capacity())

Reply:
`transfer_to_old` does check if young has enough affiliated regions to transfer:
```
  if (young_gen->free_unaffiliated_regions() < regions) {
    return false;
```

I think the original error here was caused by the Mutator free set holding many more regions than were available to the young generation (the issue with `reserve_regions`). I don't expect that to happen again, but I did add a log message if the transfer fails.

We are already somewhat loose when it comes to enforcing generational capacities. Are you proposing we have `shFreeSet` be the sole arbiter of capacity? That sounds like a good simplification - also sounds like a different PR?

Reply:
I understand that it checks, but the check should never fail, because the two asserts that precede the call (and did not need to be removed) confirm that the region to be flipped is in the Mutator free set (i.e. is part of young capacity) and is entirely empty (so must be unaffiliated).

I need to study the reserve_regions problem more.  I saw that you refactored a bit.  I didn't catch that you were fixing a bug.

Reply:
Agree that changing shFreeSet to sole arbiter is probably a separate PR, that might be a good improvement.

Reply:
Given that full-gc is allowed to violate the min/max constraints on generation sizes, I'm inclined to allow the same here, and revert this to force_transfer_to_old().  If that doesn't work, maybe a comment to explain why.  Because I'm not understanding this entirely.

Reply:
Yes, there was a logic error in `rebuild` that caused the loop to exit well before any regions were transferred to the collector's reserves. This left it in a state where `capacity_of(Mutator)` was well higher than it ought to have been (i.e., when `young->available() == 0`).